core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -257,6 +265,25 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  // Changes are initially deferrable when using a Raft-based metadata quorum, and may flip-flop thereafter;
+  // changes are never deferrable when using ZooKeeper.  When true, this indicates that we should transition online
+  // partitions to the deferred state if we see metadata with a different leader epoch.
+  @volatile private var changesDeferrable: Boolean = usingRaftMetadataQuorum
+  stateChangeLogger.info(s"Metadata changes deferrable=$changesDeferrable")
+
+  private def confirmChangesNotDeferrableWithZooKeeper(): Unit = {
+    if (changesDeferrable) {
+      throw new IllegalStateException("Partition metadata changes should never be deferrable when using ZooKeeper")
+    }
+  }
+
+  def deferrableMetadataChanges(): Unit = {
+    replicaStateChangeLock synchronized {
```

Comment:
Is it the case that whenever we grab this lock we want to assert that we're in a legal state wrt ZK and `changesDeferrable`? If that's the case, we might want to make a note of it where we define `replicaStateChangeLock`. 

Alternatively, maybe we could put this behind some helper function that grabs the lock and checks our state.

Reply:
It isn't entirely clear since there are places where this lock is acquired that we won't implement for the initial version of a Raft-based metadata quorum (e.g. handling log dir failure).  I think it's a good point that we should check here, though, because we must never make changes deferrable if we are using ZooKeeper.  So I added the check.